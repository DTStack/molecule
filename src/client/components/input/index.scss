@import '../../classNames/style.scss';

$container: prefix('textarea');
$widget: BEMElement($container, 'widget');
$validation: BEMElement($container, 'validation');

$info: BEMModifier($container, 'info');
$warning: BEMModifier($container, 'warning');
$error: BEMModifier($container, 'error');
$validating: BEMModifier($container, 'validating');

$small: BEMModifier($container, 'small');

$widget-size: 26px;
$small-widget-size: 20px;

/* prettier-ignore */
$status-colors: (
    $info: (border: var(--inputValidation-infoBorder), background: var(--inputValidation-infoBackground)),
    $warning: (border: var(--inputValidation-warningBorder), background: var(--inputValidation-warningBackground)),
    $error: (border: var(--inputValidation-errorBorder), background: var(--inputValidation-errorBackground)),
    $validating: (border: var(--inputValidation-infoBorder), background: var(--inputValidation-infoBackground)),
);

@mixin status($border, $background) {
    .#{$widget} {
        outline: 1px solid $border;
        outline-offset: -1px;
    }

    .#{$validation} {
        background: $background;
        border: 1px solid $border;
    }
}

.#{$container} {
    position: relative;
    font-size: 14px;
    line-height: 0;

    &.#{$small} {
        font-size: 13px;

        .#{$widget} {
            padding: 2px 0 2px 1px;
            height: $small-widget-size;
            min-height: $small-widget-size;
            line-height: calc($small-widget-size - 4px);
        }

        .#{$validation} {
            top: calc($small-widget-size - 1px);
        }
    }

    .#{$widget} {
        background: var(--input-background);
        border: 1px solid var(--input-border, transparent);
        box-sizing: border-box;
        color: var(--input-foreground);
        font-family: inherit;
        font-size: inherit;
        height: $widget-size;
        min-height: $widget-size;
        overflow: hidden;
        padding: 4px;
        resize: none;
        width: 100%;

        // TODO: better to mock a scroll bar
        // invisible scroll bar
        &::-webkit-scrollbar {
            opacity: 0;
        }
    }

    .#{$widget}:focus ~ .#{$validation} {
        display: block;
    }

    .#{$validation} {
        padding: 4px;
        position: absolute;
        z-index: 99;
        top: calc($widget-size - 1px);
        width: 100%;
        box-sizing: border-box;
        display: none;
        line-height: 1.15;
    }

    @each $name, $status in $status-colors {
        &.#{$name} {
            @include status(map-get($status, 'border'), map-get($status, 'background'));
        }
    }
}

:export {
    container: $container;
    small: $small;
    widget: $widget;
    validation: $validation;
    info: $info;
    warning: $warning;
    error: $error;
    validating: $validating;
    widgetSize: $widget-size;
    smallWidgetSize: $small-widget-size;
}
