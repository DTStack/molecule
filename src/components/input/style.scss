@import 'mo/style/common';

#{$input} {
    background-color: var(--input-background);
    border: 1px solid transparent;
    border-color: var(--input-border);
    color: var(--input-foreground);
    display: inline-block;
    font-size: 13px;
    min-width: 0;
    padding: 4px 11px;
    position: relative;
    transition: all 0.3s;
    width: calc(100% - 16px);

    &--lg {
        font-size: 16px;
        height: 40px;
        padding: 6px 11px;
    }

    &--disabled {
        background-color: var(--input-background);
        color: var(--input-foreground);
        cursor: not-allowed;
        opacity: 1;
    }

    &__textarea {
        &--show-count::after {
            color: var(--input-foreground);
            content: attr(data-count);
            display: block;
            text-align: right;
        }
    }

    &::placeholder {
        color: var(--input-placeholderForeground);
    }

    &:focus,
    &:active {
        border-color: var(--focusBorder);
        outline-color: var(--focusBorder);
        outline-offset: -1;
        outline-style: solid;
    }
}
