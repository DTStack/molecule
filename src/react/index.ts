export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
