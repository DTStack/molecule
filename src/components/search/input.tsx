import React, { useState, useRef } from 'react';
import { ActionBar, IActionBarItemProps } from 'mo/components/actionBar';
import {
    inputGroupClassName,
    searchToolBarClassName,
    validationBaseInputClassName,
    validationErrorInputClassName,
    validationInfoInputClassName,
    validationWarningInputClassName,
} from './base';
import { classNames } from 'mo/common/className';

export enum InfoTypeEnums {
    info = 'info',
    warning = 'warning',
    error = 'error',
}
export type InfoTypeEnum = keyof typeof InfoTypeEnums;

export interface IBaseInputProps {
    value?: string;
    className?: string;
    placeholder?: string;
    toolbarData?: IActionBarItemProps[];
    info?: { type: InfoTypeEnum; text: string };
    onChange?: (value: string) => void;
    onToolbarClick?: (addon) => void;
}

/**
 * Mock an Input by textarea
 * 'Cause we have to achieve text wrap and input cannot achieve it
 */
function Input(props: IBaseInputProps) {
    const {
        className,
        placeholder,
        toolbarData = [],
        onChange,
        value,
        info,
    } = props;

    const [focusStatus, setFocus] = useState(false);
    const textareaRef = useRef<HTMLTextAreaElement>(null);

    const onToolbarClick = (e, item) => {
        // toolbar click can trigger input focus
        textareaRef.current?.focus();
        props.onToolbarClick?.(item);
    };

    const getInfoClassName = (classname: string) => {
        switch (classname) {
            case InfoTypeEnums.info:
                return validationInfoInputClassName;
            case InfoTypeEnums.warning:
                return validationWarningInputClassName;
            case InfoTypeEnums.error:
                return validationErrorInputClassName;
            default:
                return '';
        }
    };

    const handleInputFocus = () => {
        setFocus(true);
    };

    const handleInputBlur = () => {
        setFocus(false);
    };

    const handleInputChange = (e) => {
        if (textareaRef.current) {
            // base height
            textareaRef.current.style.height = '24px';
            const curretnScollerHeight = textareaRef.current.scrollHeight;
            // count the lines
            const lines = curretnScollerHeight / 24;
            const maxLines = 5;
            if (lines > maxLines) {
                textareaRef.current.style.height = `${24 * maxLines}px`;
            } else {
                textareaRef.current.style.height = `${textareaRef.current.scrollHeight}px`;
            }
        }
        onChange?.(e.target.value || '');
    };

    const handleInputKeyPress = (e) => {
        // detect Enter press
        if (e.keyCode === 13) {
            onChange?.(e.target.value || '');
            e.preventDefault();
        }
    };

    return (
        <div className={className}>
            <textarea
                ref={textareaRef}
                spellCheck={false}
                autoCorrect="off"
                autoCapitalize="off"
                className={classNames(
                    info?.text && getInfoClassName(info?.type || '')
                )}
                value={value || ''}
                placeholder={placeholder}
                title={placeholder}
                onKeyDown={handleInputKeyPress}
                onFocus={handleInputFocus}
                onBlur={handleInputBlur}
                onChange={handleInputChange}
            />
            {info?.text && focusStatus && (
                <div
                    className={classNames(
                        validationBaseInputClassName,
                        getInfoClassName(info.type)
                    )}
                >
                    {info.text}
                </div>
            )}
            <ActionBar
                className={searchToolBarClassName}
                data={toolbarData}
                onClick={onToolbarClick}
            />
        </div>
    );
}

function Group({ children }) {
    return <div className={inputGroupClassName}>{children}</div>;
}

Input.Group = Group;

export default Input;
