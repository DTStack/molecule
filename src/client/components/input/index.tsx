import { useLayoutEffect, useRef } from 'react';
import { classNames } from 'mo/client/classNames';
import { InputValidateInfo, ValidateStatus, ValidateStatusLiteral } from 'mo/types';

import Display from '../display';
import variables from './index.scss';

export interface IInputProps {
    value?: string;
    className?: string;
    placeholder?: string;
    info?: InputValidateInfo;
    defaultValue?: string;
    autoFocus?: boolean;
    size?: 'default' | 'small';
    autoWrap?: boolean;
    onChange?: (value: string) => void;
    onClick?: React.MouseEventHandler<HTMLTextAreaElement>;
    onKeyDown?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    onSubmit?: (value: string) => void;
    onBlur?: React.FocusEventHandler<HTMLTextAreaElement>;
}

export default function Input({
    defaultValue,
    size,
    value,
    className,
    placeholder,
    autoFocus,
    autoWrap,
    info,
    onChange,
    onSubmit,
    onBlur,
    onClick,
    onKeyDown,
}: IInputProps) {
    const textareaRef = useRef<HTMLTextAreaElement>(null);

    const getInputClassName = (status: ValidateStatusLiteral) => {
        switch (status) {
            case ValidateStatus.info:
                return variables.info;
            case ValidateStatus.warning:
                return variables.warning;
            case ValidateStatus.error:
                return variables.error;
            case ValidateStatus.validating:
                return variables.validating;
            default:
                return '';
        }
    };

    const handleInputBlur = (e: React.FocusEvent<HTMLTextAreaElement, Element>) => {
        e.stopPropagation();
        e.persist();
        onBlur?.(e);
    };

    const handleInputChange = (e: any) => {
        const widgetSize = size === 'small' ? variables.smallWidgetSize : variables.widgetSize;
        const height = parseInt(widgetSize, 10);
        if (textareaRef.current) {
            // base height
            textareaRef.current.style.height = widgetSize;
            const currentScrollHeight = textareaRef.current.scrollHeight;
            // count the lines
            const lines = currentScrollHeight / height;
            const maxLines = 5;
            if (lines > maxLines) {
                textareaRef.current.style.height = `${height * maxLines}px`;
            } else {
                textareaRef.current.style.height = `${textareaRef.current.scrollHeight}px`;
            }
        }
        onChange?.(e?.target?.value || '');
    };

    const handleInputKeyPress = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        e.stopPropagation();
        e.persist();
        onKeyDown?.(e);
        // enter press
        if (e.keyCode === 13) {
            onSubmit?.((e.target as HTMLTextAreaElement).value);
            e.preventDefault();
        }
    };

    useLayoutEffect(() => {
        if (autoFocus) {
            if (textareaRef.current) {
                textareaRef.current.focus();
                const ext = ((defaultValue || '') as string).lastIndexOf('.');
                textareaRef.current.selectionStart = 0;
                textareaRef.current.selectionEnd =
                    // if period at position of 0, then this period means hidden file
                    ext > 0 ? ext : ((defaultValue || '') as string).length;
            }
        }
    }, []);

    return (
        <div
            className={classNames(
                variables.container,
                info?.message && getInputClassName(info?.status || ValidateStatus.info),
                size === 'small' && variables.small,
                className
            )}
        >
            <textarea
                defaultValue={defaultValue}
                value={value}
                ref={textareaRef}
                spellCheck={false}
                wrap={autoWrap ? 'soft' : 'off'}
                autoCorrect="off"
                autoComplete="off"
                autoCapitalize="off"
                className={classNames(variables.widget)}
                placeholder={placeholder}
                title={placeholder}
                onKeyDown={handleInputKeyPress}
                onBlur={handleInputBlur}
                onChange={handleInputChange}
                onClick={onClick}
            />
            <Display visible={Boolean(info?.message)}>
                <div className={classNames(variables.validation)}>{info?.message}</div>
            </Display>
        </div>
    );
}
