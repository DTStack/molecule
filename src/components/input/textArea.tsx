import React from 'react';
import RcTextArea, { TextAreaProps } from 'rc-textarea';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import { useEffect, useRef } from 'react';
import { omit } from 'lodash';

import { classNames, getBEMElement, getBEMModifier } from 'mo/common/className';
import { fixControlledValue, inputClassName, resolveOnChange } from './input';

export interface ITextAreaProps extends TextAreaProps {
    showCount?: boolean;
    maxLength?: number;
    onChange?: (
        e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
    ) => void;
}

const textAreaClassName = getBEMElement(inputClassName, 'textarea');
const showCountClassName = getBEMModifier(textAreaClassName, 'show-count');

export const TextArea = ({
    showCount = false,
    maxLength,
    className,
    style,
    onChange,
    ...props
}: ITextAreaProps) => {
    const innerRef = useRef(null);

    const [value, setValue] = useMergedState(props.defaultValue, {
        value: props.value,
    });

    const prevValue = useRef(props.value);

    useEffect(() => {
        if (props.value !== undefined || prevValue.current !== props.value) {
            setValue(props.value);
            prevValue.current = props.value;
        }
    }, [props.value, prevValue.current]);

    const handleSetValue = (val: string) => {
        if (props.value === undefined) {
            setValue(val);
        }
    };

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        handleSetValue(e.target.value);
        resolveOnChange(innerRef.current!, e, onChange);
    };

    const otherProps = omit(props, ['value']);

    const textArea = (
        <RcTextArea
            {...otherProps}
            value={value}
            maxLength={maxLength}
            className={classNames(className && !showCount ? [className!] : '')}
            style={showCount ? {} : style}
            prefixCls={inputClassName}
            onChange={handleChange}
            ref={innerRef}
        />
    );

    let val = fixControlledValue(value) as string;
    const hasMaxLength = Number(maxLength) > 0;

    val = hasMaxLength ? [...val].slice(0, maxLength).join('') : val;
    // Only show text area wrapper when needed
    if (showCount) {
        const valueLength = [...val].length;
        const dataCount = `${valueLength}${
            hasMaxLength ? ` / ${maxLength}` : ''
        }`;

        return (
            <div
                className={classNames(
                    className,
                    textAreaClassName,
                    showCountClassName
                )}
                style={style}
                data-count={dataCount}
            >
                {textArea}
            </div>
        );
    }
    return textArea;
};
