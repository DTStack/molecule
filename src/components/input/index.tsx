import { Input } from './input';
import { TextArea } from './textArea';

Input.TextArea = TextArea;
export default Input;
