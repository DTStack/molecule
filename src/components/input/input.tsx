import React from 'react';
import { classNames, prefixClaName, getBEMModifier } from 'mo/common/className';
import { KeyCodes } from 'mo/common/keyCodes';

import { TextArea } from './textArea';

type SizeType = 'normal' | 'large';
export interface IInputProps
    extends Omit<
        React.InputHTMLAttributes<HTMLInputElement>,
        'size' | 'onChange' | 'onKeyDown' | 'onPressEnter'
    > {
    disabled?: boolean;
    size?: SizeType;
    type?: LiteralUnion<
        'button' | 'checkbox' | 'search' | 'submit' | 'text',
        string
    >;
    placeholder?: string;
    value?: string;
    style?: React.CSSProperties;
    defaultValue?: string;
    className?: string;
    onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
    onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    onChange?: (
        e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => void;
}
export const inputClassName = prefixClaName('input');
export const normalClassName = getBEMModifier(inputClassName, 'normal');
export const largetClassName = getBEMModifier(inputClassName, 'lg');
export const disabledClassName = getBEMModifier(inputClassName, 'disabled');

export function fixControlledValue<T>(value: T) {
    if (typeof value === 'undefined' || value === null) return '';
    return value;
}

export function resolveOnChange(
    _: HTMLInputElement | HTMLTextAreaElement | null,
    e:
        | React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
        | React.MouseEvent<HTMLElement, MouseEvent>,
    onChange?: (
        event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => void
) {
    if (onChange) {
        const event = e;
        onChange(
            event as React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
        );
    }
}

export function getInputClassName(size?: SizeType, disabled?: boolean) {
    return classNames(
        inputClassName,
        { [normalClassName]: size === 'normal' },
        { [largetClassName]: size === 'large' },
        { [disabledClassName]: disabled }
    );
}

export interface InputState {
    value: any;
    prevValue: any;
}

export class Input extends React.Component<IInputProps, InputState> {
    static TextArea: typeof TextArea;

    static defaultProps = {
        type: 'text',
    };

    input: HTMLInputElement | HTMLTextAreaElement | null = null;

    constructor(props: IInputProps) {
        super(props);
        const value =
            typeof props.value === 'undefined'
                ? props.defaultValue
                : props.value;

        this.state = {
            value,
            prevValue: props.value,
        };
    }

    static getDerivedStateFromProps(
        nextProps: IInputProps,
        { prevValue }: InputState
    ) {
        const newState: Partial<InputState> = { prevValue: nextProps.value };

        if (nextProps.value !== undefined || prevValue !== nextProps.value) {
            newState.value = nextProps.value;
        }
        return newState;
    }

    saveInput = (input: HTMLInputElement) => {
        this.input = input;
    };

    setValue(value: string) {
        if (this.props.value === undefined) {
            this.setState({ value });
        }
    }

    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { onChange } = this.props;

        this.setValue(e.target.value);
        if (typeof onChange === 'function')
            resolveOnChange(this.input, e, onChange);
    };

    handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
        const { onPressEnter, onKeyDown } = this.props;

        if (e.key === KeyCodes.ENTER) {
            onPressEnter?.(e);
        }
        onKeyDown?.(e);
    };

    render() {
        const { value } = this.state;
        const {
            className,
            size = 'normal',
            disabled = false,
            placeholder,
            style,
            ...props
        } = this.props;

        return (
            <input
                value={value}
                style={style}
                placeholder={placeholder}
                onChange={this.handleChange}
                onKeyDown={this.handleKeyDown}
                className={classNames(
                    className,
                    getInputClassName(size, disabled)
                )}
                ref={this.saveInput}
                {...props}
            />
        );
    }
}
