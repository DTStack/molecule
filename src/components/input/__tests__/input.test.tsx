import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Input } from '../index';
import { disabledClassName, normalClassName, largetClassName } from '../input';

const { TextArea } = Input;
const TEST_ID = 'input';
const mockData = {
    placeholder: 'basic usage',
};

describe('Test Input Component', () => {
    test('Input snapshot', () => {
        const component = renderer.create(<Input {...mockData} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Input disabled', () => {
        const wrapper = render(
            <Input data-testid={TEST_ID} {...mockData} disabled />
        );
        const classContain = wrapper
            .getByTestId(TEST_ID)
            .classList.contains(disabledClassName);

        expect(classContain).toBeTruthy();
    });

    test('Input normal size', () => {
        const wrapper = render(<Input data-testid={TEST_ID} />);
        const classContain = wrapper
            .getByTestId(TEST_ID)
            .classList.contains(normalClassName);

        expect(classContain).toBeTruthy();
    });

    test('Input large size', () => {
        const wrapper = render(
            <Input data-testid={TEST_ID} {...mockData} size="large" />
        );
        const classContain = wrapper
            .getByTestId(TEST_ID)
            .classList.contains(largetClassName);

        expect(classContain).toBeTruthy();
    });

    test('Input type', () => {
        const INPUT_TYPE = 'search';
        const wrapper = render(
            <Input
                data-testid={TEST_ID}
                {...mockData}
                type={INPUT_TYPE}
                size="large"
            />
        );
        const typeContain = wrapper.getByTestId(TEST_ID).getAttribute('type');

        expect(typeContain).toBe(INPUT_TYPE);
    });

    test('Input placeholder', () => {
        const INPUT_PLACEHOLDER = 'basic usage';
        const wrapper = render(
            <Input data-testid={TEST_ID} placeholder={INPUT_PLACEHOLDER} />
        );
        const placeHolderContain = wrapper
            .getByTestId(TEST_ID)
            .getAttribute('placeholder');

        expect(placeHolderContain).toBe(INPUT_PLACEHOLDER);
    });

    test('Input value', () => {
        const INPUT_PLACEHOLDER = 'basic usage';
        const wrapper = render(
            <Input data-testid={TEST_ID} placeholder={INPUT_PLACEHOLDER} />
        );
        const placeHolderContain = wrapper
            .getByTestId(TEST_ID)
            .getAttribute('placeholder');

        expect(placeHolderContain).toBe(INPUT_PLACEHOLDER);
    });

    test('Input style', () => {
        const INPUT_BGCOLOR = 'red';
        const wrapper = render(
            <Input
                data-testid={TEST_ID}
                style={{ background: INPUT_BGCOLOR }}
            />
        );
        const placeHolderContain =
            wrapper.getByTestId(TEST_ID).style.background;

        expect(placeHolderContain).toBe(INPUT_BGCOLOR);
    });

    test('Input value', () => {
        const VALUE_INPUT = 'preset content';
        const wrapper = render(
            <Input data-testid={TEST_ID} value={VALUE_INPUT} />
        );
        const valueContain = wrapper.getByTestId(TEST_ID).getAttribute('value');

        expect(valueContain).toBe(VALUE_INPUT);
    });

    test('Input default value', () => {
        const DEFAULT_VALUE_INPUT = 'defaultValue';
        const wrapper = render(
            <Input data-testid={TEST_ID} defaultValue={DEFAULT_VALUE_INPUT} />
        );
        const valueContain = wrapper.getByTestId(TEST_ID).getAttribute('value');

        expect(valueContain).toBe(DEFAULT_VALUE_INPUT);
    });

    test('Input customer className', () => {
        const DEFAULT_VALUE_INPUT = 'test-className';
        const wrapper = render(
            <Input data-testid={TEST_ID} className={DEFAULT_VALUE_INPUT} />
        );
        const valueContain = wrapper
            .getByTestId(TEST_ID)
            .classList.contains(DEFAULT_VALUE_INPUT);

        expect(valueContain).toBeTruthy();
    });

    test('Input customer keyDown event', () => {
        const TEST_EVENT = jest.fn();
        const wrapper = render(
            <Input data-testid={TEST_ID} onKeyDown={TEST_EVENT} />
        );
        fireEvent.keyDown(wrapper.getByTestId(TEST_ID));

        expect(TEST_EVENT).toBeCalled();
    });

    test('Input customer pressEnter event', () => {
        const TEST_EVENT = jest.fn();
        const wrapper = render(
            <Input data-testid={TEST_ID} onPressEnter={TEST_EVENT} />
        );
        fireEvent.keyDown(wrapper.getByTestId(TEST_ID), {
            key: 'Enter',
            keyCode: 13,
        });

        expect(TEST_EVENT).toBeCalled();
    });

    test('Input customer onchange event', () => {
        const TEST_EVENT = jest.fn();
        const TEST_VALUE = '1';
        const wrapper = render(
            <Input data-testid={TEST_ID} onChange={TEST_EVENT} />
        );
        const element = wrapper.getByTestId(TEST_ID);

        fireEvent.change(element, {
            target: { value: TEST_VALUE },
        });

        expect(TEST_EVENT).toBeCalled();
    });

    test('Input built-in onchange event', () => {
        const TEST_VALUE = '1';
        const wrapper = render(<Input data-testid={TEST_ID} />);
        const element = wrapper.getByTestId(TEST_ID);

        fireEvent.change(element, {
            target: { value: TEST_VALUE },
        });

        expect(element.getAttribute('value')).toBe(TEST_VALUE);
    });

    test('Overwrite the current state value with props', () => {
        const BUTTON_ID = 'button';
        const START_VALUE = '0';
        const INPUT_VALUE = '1';
        const ENDED_VALUE = '100';
        class Demo extends React.Component {
            state = {
                value: START_VALUE,
            };

            up = () => {
                this.setState({
                    value: ENDED_VALUE,
                });
            };

            render() {
                const { value } = this.state;
                return (
                    <>
                        <Input data-testid={TEST_ID} value={value} />
                        <button
                            data-testid={BUTTON_ID}
                            type="button"
                            onClick={this.up}
                        >
                            up
                        </button>
                    </>
                );
            }
        }

        const wrapper = render(<Demo />);
        const inputElement = wrapper.getByTestId(TEST_ID);
        const buttonElement = wrapper.getByTestId(BUTTON_ID);

        expect(inputElement.getAttribute('value')).toBe(START_VALUE);

        inputElement.setAttribute('value', INPUT_VALUE);
        expect(inputElement.getAttribute('value')).toBe(INPUT_VALUE);

        fireEvent.click(buttonElement);
        expect(inputElement.getAttribute('value')).toBe('100');
    });
});

describe('Test TextArea Component', () => {
    test('TextArea snapshot', () => {
        const component = renderer.create(<TextArea {...mockData} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('TextArea showCount and maxLength', () => {
        const MAX_LENGTH = 120;
        const mockData = {
            showCount: true,
            maxLength: MAX_LENGTH,
        };

        render(<TextArea {...mockData} />);

        const count = screen.getByText(/.*/i, { selector: 'textarea' });
        const parent = count?.parentNode as any;

        expect(count.getAttribute('maxlength')).toBe(MAX_LENGTH.toString());
        expect(parent?.dataset?.count).toBe(`0 / ${MAX_LENGTH}`);
    });

    test('TextArea change showCount and maxLength', () => {
        const MAX_LENGTH = 120;
        const VALUE = 'molecule';
        const mockData = {
            showCount: true,
            maxLength: MAX_LENGTH,
            value: VALUE,
        };

        render(<TextArea {...mockData} />);

        const count = screen.getByText(/.*/i, { selector: 'textarea' });
        const parent = count?.parentNode as any;

        expect(count.getAttribute('maxlength')).toBe(MAX_LENGTH.toString());
        expect(parent?.dataset?.count).toBe(`${VALUE.length} / ${MAX_LENGTH}`);
    });

    test('TextArea change event', () => {
        const MOCK_EVENT = jest.fn();
        const TEST_VALUE = 'test value';
        const wrapper = render(
            <TextArea data-testid={TEST_ID} onChange={MOCK_EVENT} />
        );
        const component = wrapper.getByTestId(TEST_ID);

        fireEvent.change(component, { target: { value: TEST_VALUE } });

        expect(MOCK_EVENT).toBeCalled();
    });
});
