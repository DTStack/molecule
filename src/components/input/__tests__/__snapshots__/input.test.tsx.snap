// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Input Component Input snapshot 1`] = `
<input
  className="mo-input mo-input--normal"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="basic usage"
  type="text"
/>
`;

exports[`Test TextArea Component TextArea snapshot 1`] = `
<textarea
  className="mo-input"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="basic usage"
  style={Object {}}
  value=""
/>
`;
